codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/StructureGenerator.java

```diff
@@ -149,8 +157,18 @@ private void renderErrorStructure() {
             });
             writer.write("return *e.Message");
         });
-        writer.write("func (e *$L) ErrorCode() string { return $S }",
-                structureSymbol.getName(), shape.getId().getName(service));
+
+        writer.openBlock("func (e *$L) ErrorCode() string {", "}", structureSymbol.getName(), () -> {
+            writer.write("errorCode := $S", shape.getId().getName(service));
+
+            // override error code via integration if present
+            for (GoIntegration integration: integrations) {
+                integration.onShapeWriterUse(settings, model, symbolProvider, writer, shape);
```

Comment:
I don't think that's the intent of this hook. (Actually it looks like it's not used for its intended purpose since it isn't actually called when a shapewriter is created.)

Reply:
Could you add some more info on what the intent is then?  Also should be have a similar method on GoIntegration for this use-case where I want to generate code for a specific shape?

Reply:
The intent is to add a hook for whenever a shape writer is initially used, for whatever reason you might care to think of. You should use a method that is more purpose-specific. You could just rename this method to something more appropriate, since it's never actually used.

A better approach would be to have a `getErrorCode` or some such on protocol generators where the default implementation is the previous behavior. Then you would override that in the query protocols.

Reply:
Updated to use `getErrorCode` method on protocol generator.